RepositorySimulator: add non-consistent snapshot support

sechkova opened the pull request:
Extend filename partitioning to support serving non-versioned
metadata and non-prefixed target files.

Fixes #1637, fixes #1667, fixes #1573

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

sechkova committed:
RepositorySimulator: add non-consistent snapshot support

Extend URL partitioning to support serving non-versioned
metadata and non-prefixed target files.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
I didn't add tests in this PR, I opened this one instead: #1667

jku commented on tests/repository_simulator.py:
```diff
@@ -185,24 +200,19 @@ def publish_root(self):
-                role = "timestamp"
-            else:
-                version, _, role = ver_and_name.partition(".")
+            role, version = self._partition_filename(ver_and_name)
```
hmm, it seems to do the right thing, but your function is called "partition filename" and then you need to feed it something that is explicitly just a portion of a filename ...

jku reviewed (commented):
Did you somehow test this privately?

The core difficulty with this issue is the corner we painted ourselves in by telling warehouse they can be consistent_snapshot **and** offer non-consistent targets for download at the same time: so we can't 100% tell if a targetname is hash prefixed or not when a client requests it ...

This code seems to assume that _if target name contains a "." then it contains a prefix_. I realize there probably aren't perfect solutions here but this seems non-optimal: it should be possible to request non-prefixed targets with dots in their names

sechkova commented:
> This code seems to assume that _if target name contains a "." then it contains a prefix_. I realize there probably aren't perfect solutions here but this seems non-optimal: it should be possible to request non-prefixed targets with dots in their names

Oh yes, you are right. I guess I will rework it in parallel with tests to cover all the weird cases.

jku reviewed (changes requested):
marking "request changes" so it's visible in issue list

sechkova committed:
ngtests: Add consistent_snapshot tests

Add tests for ngclient.Updater toggling
'consitent_snapshot' and 'prefix_targets_with_hash'.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
I reworked the implementation and now `RepositorySimulator` makes the decision whether the client requests versioned/prefixed files or not based on: `self.root.consistent_snapshot` and `self.prefix_targets_with_hash`.  While the specification describes how repository and client sync on consistent_snapshot, I presumed that they both need to know in advance whether they will work with hash-prefixed targets and I added the same `prefix_targets_with_hash` option to the repository. 

I also added one new test file for consistent snapshot. I couldn't decide if the class `TestConsistentSnapshot` needs to be in a separate test file or integrated in another. Let me know your opinion on this as well as the general idea/readability of the tests.
Since we want to test the client and not that much the `RepositorySimulator` and it is not possible to verify that internally Updater builds the correct prefix+name, I "wrapped" `RepositorySimulator._fetch_*` to check the calls made by Updater.

sechkova commented on tests/repository_simulator.py:
```diff
@@ -192,24 +195,32 @@ def fetch(self, url: str) -> Iterator[bytes]:
-                role = "timestamp"
-            else:
+            # root is always version-prefixed while timestamp is always NOT
+            if ver_and_name.endswith("root") or (
```
This will fail if a delegated role name ends with "root" (and consistent_snapshot is not enabled). But otherwise it is hard to generalize and I guess we don't want to support that weird cases in RepositorySimulator?

jku replied:
yeah no need to spend a lot of time polishing a test utility to perfection... 

that said you could move `version, _, role = ver_and_name.partition(".")` to before this line so you could compare `role == "root"` (but would keep rest of the code as is -- I admit the else clause is unintuitive if this is done)?

up to you

sechkova replied:
I tried to do it in a readable way ... at least now it is correct for the cases I can think of.

jku commented on tests/test_updater_consistent_snapshot.py:
```diff
@@ -0,0 +1,236 @@
+        for targetpath in targetpaths:
+            sim.targets.version += 1
+            sim.add_target("targets", b"content", targetpath)
+            sim.update_snapshot()
```
It's not important but there should be no need to update snapshot more than once

jku commented on tests/test_updater_consistent_snapshot.py:
```diff
@@ -0,0 +1,236 @@
+# Copyright 2021, New York University and the TUF contributors
+# SPDX-License-Identifier: MIT OR Apache-2.0
+
+"""Test ngclient Updater consistent snapshot"""
```
this could be better (we're testing consistent snapshot in all client tests, hopefully). This file seems to test 
* that the client works with all combinations of consistent_snapshot and prefix_targets_with_hash
* that expected URLs are being fetched with all of these combinations

jku reviewed (commented):
I think your reasoning about prefix_targets_with_hash option makes sense.

I also think counting the fetches at repository side is fine and actually better than poking into updater internals. 

The patching in the tests is not pretty but it is much improved in this version. I think the asserts you use do not check whether _other_ fetches were also made though, which I think we want (so we know it doesn't download things twice or something)... I think we should be comparing our expected calls list directly to the actually made calls list. WDYT?

...which leads me to: Making the call counters a RepositorySimulator feature might make the tests more readable and probably wouldn't be a lot of code: `sim.call_stats.metadata` could contain basically the same list of tuples that your testdata contains -- but this could also be a followup idea if you'd rather get these tests in ASAP.

sechkova commented:
> ...which leads me to: Making the call counters a RepositorySimulator feature might make the tests more readable and probably wouldn't be a lot of code: sim.call_stats.metadata could contain basically the same list of tuples that your testdata contains -- but this could also be a followup idea if you'd rather get these tests in ASAP.

I think I will open a separate issue. It may double some of the work but it would be easier at least for me to do it as a follow up.
There it is: #1682

sechkova commented:
Just noting that I'd like to squash the last two commits before merging :)

jku reviewed (approved):
:+1:

sechkova commented:
Squashed the last two commits, should be ready now.

sechkova commented:
I now saw this issue:  #1573
I think `test_refresh_on_consistent_targets()` can be replaced by the new test file (`test_updater_consistent_snapshot.py`) and we can close it. I can do it in this PR if you agree?

jku commented:
Sounds good to me!

sechkova committed:
Remove test_refresh_on_consistent_targets

Consistent snapshot and consistent targets are now
extensively tested in test_updater_consistent_snapshot.py.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

jku commented:
Thanks
